src/RawPOSIX/src/interface/signal.rs

```diff
@@ -0,0 +1,245 @@
+use crate::{constants::{SA_NODEFER, SA_RESETHAND, SIG_DFL}, interface::{cagetable_getref, cagetable_getref_opt, RustAtomicOrdering}};
+
+const EPOCH_NORMAL: u64 = 0;
+const EPOCH_SIGNAL: u64 = 0xc0ffee;
+const EPOCH_KILLED: u64 = 0xdead;
+
+// switch the epoch of the main thread of the cage to "signal" state
+pub fn signal_epoch_trigger(cageid: u64) {
+    let cage = cagetable_getref(cageid);
+    let main_threadid = cage.main_threadid.load(RustAtomicOrdering::Relaxed) as i32;
+    let epoch_handler = cage.epoch_handler.get(&main_threadid).expect("main threadid does not exist");
+    let guard = epoch_handler.write();
+    let epoch = *guard;
+    // SAFETY: the pointer is locked with write access so no one is able to modify it concurrently
+    // However, Potential BUG (TODO): We still need to verify the lifetime of the pointer. This pointer
+    // is created by wasmtime and will be destroyed at some point when the wasm instance is destroyed
+    // we still need to figure out when is the destroy happening and make sure it is destroyed after the
+    // information in rawposix is updated
+    unsafe {
+        *epoch = EPOCH_SIGNAL;
+    }
+}
+
+// switch the epoch of all threads of the cage to "killed" state
+pub fn epoch_kill_all(cageid: u64) {
+    let cage = cagetable_getref(cageid);
+    let main_threadid = cage.main_threadid.load(RustAtomicOrdering::Relaxed) as i32;
+    // we iterate through the epoch handler of each thread in the cage
+    for entry in cage.epoch_handler.iter() {
+        if entry.key() == &main_threadid {
+            // main thread should be the one invoking this method
+            // so main thread could kill itself and we do not need to notify it again
+            continue;
+        }
+        let epoch_handler = entry.value();
+        let guard = epoch_handler.write();
+        let epoch = *guard;
+        // SAFETY: see comment at `signal_epoch_trigger`
+        unsafe {
+            *epoch = EPOCH_KILLED;
+        }
+    }
+}
+
+// check the specified thread with specified cage is in "killed" state
+pub fn thread_check_killed(cageid: u64, thread_id: u64) -> bool {
+    let cage = cagetable_getref(cageid);
+    let epoch_handler = cage.epoch_handler.get(&(thread_id as i32)).unwrap();
+    let guard = epoch_handler.write();
+    let epoch = *guard;
+    // SAFETY: see comment at `signal_epoch_trigger`
```

Comment:
I like this style.   Good job commenting the code to explain what is happening / what may not work.